### PR TITLE
More FollowApi

### PR DESCRIPTION
**Pull request by akshays2112**
Hi Daniel,

Will need to triple check if I did this right.

Everything works though but I know that is never enough.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

### CODE REVIEW

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Please change to `ArgumentException` with message "A minimum of 1 and a maximum of 50 ids can be sent."

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Just call `Put` here, e.g.

    await Put(builder.Uri, null, accessToken);

If you pull from origin master I have added a Uri overload for this helper 🙂

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Please change all range checks to ArgumentException as per previous comment

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Just call `Put` here 🙂

File: src/SpotifyApi.NetCore/Http/RestHttpClient.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/Http/RestHttpClient.cs]: These methods are not required 🙂

File: src/SpotifyApi.NetCore/SpotifyWebApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/SpotifyWebApi.cs]: This method is not required. It already exists as `Put` 🙂

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]
Apologies @akshays2112 I have a very busy weekend. I will review when I can.

Thanks for adding this new endpoint. There is already a helper for `Put` - see my comments in code. Apart from that just a minor change to the way you are validating id length. Looks really good!

Thanks for all of your help 😃 Keep it up! 👏

**Comment by akshays2112** [comment]
Hi Daniel,

Check it out when you can. Everything should be good.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by akshays2112** [comment]
Hi Daniel,

FollowApi all endpoints wrapped. Waiting on your approval before I go to the next set of endpoints.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by DanielLarsenNZ** [comment]
Hi @akshays2112 , I’m probably not going to be able to review this for 36 hours sorry. You are welcome to keep wrapping endpoints. You can continue on this PR or create a new one. Up to you 😀

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: For this validation check the `ArgumentNullException` is correct. I prefer `string.IsNullOrEmpty()` (it is more efficient and easier to read). It may be a minor thing, but you should avoid counting the length of a string unless you absolutely have to 🙂 e.g.

    if (string.IsNullOrEmpty(playlistId)) throw new ArgumentNullException("playlistId");

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs:135]: This is good - thank you 😃

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: This should default to `null` and be consistent with the generic method

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: Use `builder.AppendToQueryIfValueNotNullOrWhiteSpace()` here, e.g.

    builder.AppendToQueryIfValueNotNullOrWhiteSpace("after", after);

File: src/SpotifyApi.NetCore/FollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/FollowApi.cs]: See comment above. Should be:

    if (string.IsNullOrEmpty(playlistId)) throw new ArgumentNullException("playlistId");

File: src/SpotifyApi.NetCore/SpotifyWebApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/SpotifyWebApi.cs:168]: Love it! 🤩

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]
Hey @akshays2112 - thanks for adding the extra endpoints. That brings us to 100% coverage of the Follow API which is fantastic. Just a few minor changes.

Also it looks like there is a merge conflict with one of my changes in SpotifyWebApi that you will need to resolve, but let me know if you need a hand fixing that.

**Comment by akshays2112** [comment]
Hi Daniel,

Made changes to all FollowApi and today's full coverage of LibraryApi added from 0 endpoints covered. I took in all your comments and corrected both previous days FollowApi and today's LibraryApi.

Asking for a small amount of trust that you can totally undo as you see and fit when you have time. Just accept my merge conflicts. Let my changes overwrite yours.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by akshays2112** [comment]
Hi Daniel,

Stopping here. All that is left is PlayerApi and PlaylistsApi. PlayerApi is all BETA so need to talk to you. PlaylistsApi if you give me permission I will reimplement it with all of your corrections to my code.

Thanks for your time and patience and time.

Best Regards,

Akshay Srinivasan

**Comment by akshays2112** [comment]
Hi Daniel,

The only merge conflicts that are not mine are in SpotifyWebApi.cs. These are required so it would be nice if you can accept the merge conflicts allowing my changes to overwrite yours.

If we can connect given the time difference and your job responsibilities then I need your help to figure on voice on how to pull from your master to merge your changes into my old fork of your code.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Comment by DanielLarsenNZ** [comment]
Hi @akshays2112,

Thank you for starting on the new endpoints! 😃 Please move those files to new branches and new PR's for each group of endpoints, i.e. one for LibraryApi, one for EpisodesApi and so on. This is so that we can keep the PR's reasonably small and release each one separately.

I will help you out with the merge conflicts. The Conflicts have to be resolved before your PR is merged into master 🙂

File: src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/EpisodeApiTests.cs:1]: Awesome! 👏 Move this file to a new branch and PR called "Episodes API" 👍

File: src/SpotifyApi.NetCore.Tests/FollowApiTests.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/FollowApiTests.cs:26]: The convention in this project is to use `var` whenever the type is obvious. See other classes for guidelines.

File: src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/PersonalizationApiTests.cs:1]: Please move to separate branch and PR "Personalization API"

File: src/SpotifyApi.NetCore.Tests/ShowsApiTests.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore.Tests/ShowsApiTests.cs:1]: Please move to separate branch and PR named "Shows API"

File: src/SpotifyApi.NetCore/EpisodesApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/EpisodesApi.cs:1]: Please move to new branch and PR named "Episodes API"

File: src/SpotifyApi.NetCore/IEpisodesApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IEpisodesApi.cs:1]: Please move to new branch and PR named "Episodes API"

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:22]: Nice one 😎

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:48]: Love it! 🤩

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:83]: Please keep the same text as the Spotify docs, i.e. "Check to see if one or more Spotify users are following a specified playlist."

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:100]: Please keep same text as Spotify docs

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:112]: Please keep same text as Spotify docs

File: src/SpotifyApi.NetCore/IFollowApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IFollowApi.cs:126]: Please keep same text as Spotify docs

File: src/SpotifyApi.NetCore/ILibraryApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/ILibraryApi.cs:1]: Please move to new branch and PR named "Library API"

File: src/SpotifyApi.NetCore/IPersonalizationApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IPersonalizationApi.cs:1]: Please move to new branch and PR named "Personalization API"

File: src/SpotifyApi.NetCore/IShowsApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/IShowsApi.cs:1]: Please move to new branch and PR named "Shows API"

File: src/SpotifyApi.NetCore/Models/Common.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/Models/Common.cs:149]: Please move to new branch and PR named "Episodes API"

File: src/SpotifyApi.NetCore/Models/Episode.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/Models/Episode.cs:1]: Please move to new branch and PR named "Episodes API"

File: src/SpotifyApi.NetCore/PersonalizationApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/PersonalizationApi.cs:1]: Please move to new branch and PR named "Personalization API"

File: src/SpotifyApi.NetCore/ShowsApi.cs
Comment (DanielLarsenNZ) [inline src/SpotifyApi.NetCore/ShowsApi.cs:1]: Please move to new branch and PR named "Shows API"

**Review by DanielLarsenNZ — CHANGES_REQUESTED** [review]

**Comment by akshays2112** [comment]
Hi Daniel,

Will do. Working on it.

Thanks for your patience and time.

Best Regards,

Akshay Srinivasan

**Pull request closed without merging**
